### DISCUSSION TITLE
Share rusqlite connection between threads with r2d2

### DISCUSSION THREAD

**Discussion by mawilms** [rusqlite/rusqlite#1183] [Questions]
Hello everyone,

for a iced GUI application I want to have a single struct which contains the sqlite connection. Some calls are async so I have to wrap the connection in an `Arc` smart pointer.

My shortened code looks like this:

```rust
#[derive(Debug, Clone)]
pub struct Cache {
    connection: Arc<PooledConnection<SqliteConnectionManager>>,
}

impl Cache {
    pub fn new(connection: PooledConnection<SqliteConnectionManager>) -> Self {
        Self {
            connection: Arc::new(connection),
        }
    }
}

pub async fn init_application() -> State {

    let manager = SqliteConnectionManager::file(&database_path);
    let pool = r2d2::Pool::new(manager).expect("Error while creating a database pool");
    let connection = pool.get().expect("Error while creating a pooled connection");

    let cache = Cache::new(connection);

    State::new(cache)
}

#[derive(Debug, Clone)]
pub struct State {
    cache: Arc<Cache>,
}

impl State {
    pub fn new(cache: Cache) -> Self {
        Self {
            cache: Arc::new(cache),
        }
    }
}
```

Unfortunately I get for example this error: **future cannot be sent between threads safely
within `r2d2::PooledConnection<r2d2_sqlite::SqliteConnectionManager>`, the trait `std::marker::Sync` is not implemented for `std::cell::RefCell<rusqlite::inner_connection::InnerConnection>`**

I have actually no clue why because the SqliteConnectionManager of r2d2 implements the Sync trait: https://docs.rs/r2d2_sqlite/0.20.0/r2d2_sqlite/struct.SqliteConnectionManager.html . The PooledConnection struct has also the Sync trait which should be a wrapper around the rusqlite connection: https://docs.rs/r2d2/latest/r2d2/struct.PooledConnection.html

I appreciate any help I can get :)

Thanks a lot!

**Comment by gwenn** [comment]
https://docs.rs/r2d2/latest/r2d2/struct.PooledConnection.html#impl-Sync
```
impl<M> Sync for PooledConnection<M>
where
    <M as ManageConnection>::Connection: Send + Sync, 
```
but here `M` is not `Sync`...
